slyd/slyd/splash/commands.py

```diff
@@ -35,55 +24,123 @@ def save_html(data, socket):
     stated_encoding = socket.tab.evaljs('document.characterSet')
+    if 'use_live' not in data:
+        data['use_live'] = _check_js_required(socket)
+    if data.get('use_live'):
+        sample['body'] = 'rendered_body'
+    else:
+        sample['body'] = 'original_body'
+    html = socket.tab.html()
+    sample['rendered_body'] = html
     try:
-        sample['original_body'] = _decode(socket.tab.network_manager._raw_html,
-                                          stated_encoding)
-        # XXX: Some pages only show a 301 page. Load the browser html instead
+        sample['original_body'] = _decode(
+            socket.tab.network_manager._raw_html, stated_encoding)
+        # XXX: Some pages only show a 301 page. Load the browser html
         assert len(sample['original_body']) > 500
     except (AttributeError, AssertionError):
-        sample['original_body'] = socket.tab.html()
+        sample['original_body'] = html
     _update_sample(data, socket, sample, save=True)
 
 
 def extract_items(data, socket):
     """Use latest annotations to extract items from current page"""
-    if not socket.tab:
+    if not socket.tab or not socket.spider or not data.get('sample'):
         return {}
     url = socket.tab.evaljs('location.href')
     html = socket.tab.html()
+    try:
+        raw_html = socket.spiderspec.spider['original_body']
+    except (TypeError, KeyError):
+        stated_encoding = socket.tab.evaljs('document.characterSet')
+        raw_html = _decode(
+            socket.tab.network_manager._raw_html, stated_encoding)
     if (socket.spiderspec is None or
             (data['spider'] and socket.spiderspec.name != data['spider'])):
         result = socket.open_spider(data)
         if result and result.get('error'):
             return {}
-    sample_names = socket.spiderspec.templates
-    annotations = socket.spider.plugins['Annotations']
-    sid = data.get('sample')
-    if sid:
-        sample_names = [_update_sample(data, socket, use_live=True)]
-        spider = {k: sample_names if k == 'templates' else v
-                  for k, v in socket.spiderspec.spider.items()}
-        items, extractors = _load_items_and_extractors(data, socket)
-        extraction = BotAnnotations()
-        extraction.setup_bot(Settings(), spider, items, extractors,
-                             _SPIDER_LOG)
-        socket.spider.plugins['Annotations'] = extraction
-    items, links = extract_data(url, html, socket.spider, sample_names)
-    if not items:
-        items, links = extract_data(url, socket.tab.network_manager._raw_html,
-                                    socket.spider, sample_names)
-    socket.spider.plugins['Annotations'] = annotations
+    schemas, extractors = _load_items_and_extractors(data, socket)
+    using_js = _check_js_required(socket, url)
+    with _restore(socket.spider):
+        js_live_items, js_raw_items, links = _load_items(
+            data, socket, schemas, extractors, url, html, raw_html,
```

Comment:
To avoid these long parameter lists, we can use a `Parameter Object` which benefits both the `extract_items` and `save_html` functions. Here is a untested first approach:

```
  class DataClump(object):
      def __init__(data, socket):
          self.data = data
          self.socket = socket
          self.spider = socker.spider
  
          self.url = socket.tab.evaljs('location.href')
          self.html = socket.tab.html()
          self.schemas, self.extractors = _load_items_and_extractors(data, socket)
  
      @property
      def using_js():
          request = Request(self.url)
          return 'splash' in self.socket.spider._add_splash_meta(request).meta
  
      @property
      def original_body():
          stated_encoding = self.socket.tab.evaljs('document.characterSet')
          return _decode(self.socket.tab.network_manager._raw_html, stated_encoding)
  
      @property
      def raw_html():
          try:
              return self.socket.spiderspec.spider['original_body']
          except (TypeError, KeyError):
              self.original_body
  
      @property
      def copy_spider():
          return self.socket.spiderspec.spider.copy()
  
      def extraction_for(spider):
          extraction = BotAnnotations()
          return extraction.setup_bot(Settings(), spider,
                                      self.schemas, self.extractors)

def save_html(data, socket):
    manager = socket.manager
    path = [s.encode('utf-8') for s in (data['spider'], data['sample'])]
    sample = _load_sample(manager, *path)
    if sample.get('rendered_body') and not data.get('update'):
        return

    data_clump = DataClump(data, socket)
    if 'use_live' not in data:
        data['use_live'] = data_clump.using_js

    body_type = 'rendered' if data.get('use_live') else 'original'
    sample['body'] = '{}_body'.format(body_type)
    sample['rendered_body'] = data_clump.html

    try:
        sample['original_body'] = data_clump.original_body
        # XXX: Some pages only show a 301 page. Load the browser html
        assert len(sample['original_body']) > 500
    except (AttributeError, AssertionError):
        sample['original_body'] = data_clump.html

    _update_sample(data, socket, sample, save=True)

def extract_items(data, socket):
    """Use latest annotations to extract items from current page"""
    if not socket.tab or not socket.spider or not data.get('sample'):
        return {}
    if (socket.spiderspec is None or
            (data['spider'] and socket.spiderspec.name != data['spider'])):
        result = socket.open_spider(data)
        if result and result.get('error'):
            return {}

    data_clump = DataClump(data, socket)

    with _restore(socket.spider):
        js_live_items, js_raw_items, links = _load_items(data_clump, use_live=True)
        live_items, raw_items, links = _load_items(data_clump, use_live=False)
        if data_clump.using_js:
            changes, changed_values = _compare_items(js_live_items, raw_items)
            sample_change, _ = _compare_items(js_live_items, live_items)
            items = js_live_items
        else:
            changes, changed_values = _compare_items(raw_items, js_raw_items)
            changes.extend(_compare_items(live_items, js_live_items)[0])
            sample_change, _ = _compare_items(raw_items, js_raw_items)
            items = raw_items if raw_items else live_items
    # TODO: add option for user to view raw and js items in UI from WS
    items = _process_items(items)
    return {'links': links, 'items': items, 'changes': changes,
            'changed': changed_values, 'type': 'js' if data_clump.using_js else 'raw'}

def _load_items(data_clump, use_live=False):
    spider = data_clump.copy_spider()
    spider['body'] = 'rendered_body' if use_live else 'original_body'
    samples = [_update_sample(data_clump.data, data_clump.socket, use_live=live)]
    spider['templates'] = samples
    socket.spider.plugins['Annotations'] = data_clump.extraction_for(spider)
    live_items, links = extract_data(data_clump.url, data_clump.html,
                                     data_clump.spider, samples)
    raw_items, links = extract_data(data_clump.url, data_clump.raw_html,
                                    data_clump.spider, samples)
    return live_items, raw_items, links

```